group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupCoordinatorService.java

```diff
@@ -443,7 +448,14 @@ public CompletableFuture<ShareGroupHeartbeatResponseData> shareGroupHeartbeat(
             topicPartitionFor(request.groupId()),
             Duration.ofMillis(config.offsetCommitTimeoutMs()),
             coordinator -> coordinator.shareGroupHeartbeat(context, request)
-        ).exceptionally(exception -> handleOperationException(
+        ).thenCompose(result -> {
+            // This ensures that the previous group write has completed successfully
+            // before we start the persister initialize phase.
+            if (result.getValue().isPresent()) {
+                return persisterInitialize(result.getValue().get(), result.getKey());
```

Comment:
My thought when writing the KIP was that, if `ShareGroupHeartbeat` needed to ask the persister to initialize one or more topic-partitions, it would schedule this work with the persister and simply 😄 omit these pending topic-partitions from the response to `ShareGroupHeartbeat`. The response should not be delayed until the persister has completed. The call to the persister involves one or more inter-broker RPCs, which could potentially need to be retried due to a cooordinator restart or load delay, so I don't think it's appropriate to do the initialization directly as part of the heartbeat processing.